### PR TITLE
FUSETOOLS-3027 - making the REST Configuration details writable

### PR DESCRIPTION
**Pull request by bfitzpat**
-- make fields on the REST tab editable for REST
Configuration details
-- enable plus button for adding a new REST Configuration to a file that
doesn't have one
-- enable X button for removing ALL REST Configuration elements & REST
Elements
-- adjusted approach to go with SWT UI hooks instead of data binding
-- added new integration test

Signed-off-by: Brian Fitzpatrick <bfitzpat@redhat.com>

# Pull Request Checklist

After this checklist is all checked or PR provides explanations for possible pass-through, please put the label "Ready for review" on the PR.

## General

- [ ] Did you use the Jira Issue number in the commit comments?
- [ ] Did you set meaningful commit comments on each commit?
- [ ] Did you sign off all commits for this PR? (git commit -s -m "jira issue number - your commit comment")

## Functional

- [ ] Did the CI job report a successful build?
- [ ] Is the non-happy path working, too?

## Maintainability

- [ ] Are all Sonar reported issues fixed or can they be ignored?
- [ ] Is there no duplicated code?
- [ ] Are method-/class-/variable-names meaningful?

## Tests

- [ ] Are there unit-tests?
- [ ] Are there integration tests (or at least a jira to tackle these)?
- [ ] Do we need a new UI test?

## Legal

- [ ] Have you used the correct file header copyright comment?
- [ ] Have you used the correct year in the headers of new files?

### CODE REVIEW

File: editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/restconfiguration/RestConfigEditor.java [hunk 1]
```diff
@@ -694,4 +807,39 @@ public void handleEvent(Event event) {
...
 	
+	/**
+	 * Public for tests.
+	 */
+	public void addRestConfigurationElement() {
+		Display.getDefault().syncExec(() -> {
+			RestConfigurationElement newrce = util.createRestConfigurationNode(ctx.getCamelFile());
+			newrce.initialize();
+			newrce.setHost("localhost");
+			newrce.setBindingMode("off");
```
[the remark this answers is not in the document]
Comment (lhein) [reply editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/restconfiguration/RestConfigEditor.java]: wondering why sonar did not complain then

File: editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/restconfiguration/RestConfigEditor.java [hunk 2]
```diff
@@ -303,6 +325,27 @@ public ImageDescriptor getImageDescriptor() {
 		}
 	}
 
+	class DeleteRestConfigurationAction extends PushAction {
+		@Override
+		public void run() {
+			MessageBox box = new MessageBox(Display.getCurrent().getActiveShell(), SWT.YES | SWT.NO);
```
[the remark this answers is not in the document]
Comment (lhein) [reply editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/restconfiguration/RestConfigEditor.java]: The port number has no validation restricting it to numbers only.

**Comment by lhein** [comment]
The port number in the rest editor has no validation which only allows numbers.

File: editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/restconfiguration/RestConfigEditor.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (lhein) [reply editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/restconfiguration/RestConfigEditor.java]: I don't like the current style of that messagebox. I would recommend using a question dialog or a warning dialog instead.

**Comment by apupier** [comment]
> The port number in the rest editor has no validation which only allows numbers.

can placeholders be used also?

File: editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/restconfiguration/RestConfigUtil.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+			}
+		}
+		return null;
+	}
+	
+	public boolean isBlueprintConfig(Node node) {
+		if (node != null) {
+			String nsURI = getNamespace(node);
+			if(!Strings.isEmpty(nsURI) && nsURI != null) {
+				return nsURI.contains("blueprint"); //$NON-NLS-1$
```
[the remark this answers is not in the document]
Comment (lhein) [reply editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/restconfiguration/RestConfigUtil.java]: not even used it seems.

**Comment by lhein** [comment]
the rest operations are not really sorted in a predictable way, no?

**Comment by lhein** [comment]
> can placeholders be used also?

I have no clue. Probably they can.

**Comment by lhein** [comment]
I addressed the here mentioned issues (I hope all of them) in this PR: https://github.com/bfitzpat/fuseide/pull/37

**Comment by bfitzpat** [comment]
Spun off validation into a separate JIRA we may address later on if we have time and it's deemed relevant - https://issues.jboss.org/browse/FUSETOOLS-3034

File: editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/restconfiguration/RestConfigUtil.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+	private static final String XMLNAMESPACE = "xmlns"; //$NON-NLS-1$
+
+	public Parameter createParameter(String name, String jType) {
+		Parameter outParm = new Parameter();
+		outParm.setName(name);
+		outParm.setJavaType(jType);
+		return outParm;
+	}
+
```
[the remark this answers is not in the document]
Comment (bfitzpat) [reply editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/restconfiguration/RestConfigUtil.java]: Removed it.

**Comment by bfitzpat** [comment]
Added separate test to verify that we are removing all REST Configuration and REST Elements successfully.

**Pull request merged**